netbox-prometheus-sd.py

```diff
@@ -6,88 +6,180 @@
 import argparse
 import itertools
 import netaddr
+import logging
+from urllib3.exceptions import RequestError
 
 import pynetbox
 
 
-def main(args):
-    targets = []
-    netbox = pynetbox.api(args.url, token=args.token)
-
-    # Filter out devices without primary IP address as it is a requirement
-    # to be polled by Prometheus
-    devices = netbox.dcim.devices.filter(has_primary_ip=True)
-    vm = netbox.virtualization.virtual_machines.filter(has_primary_ip=True)
-    ips = netbox.ipam.ip_addresses.filter(**{'cf_%s' % args.custom_field: '{'})
-
-    for device in itertools.chain(devices, vm, ips):
-        if device.custom_fields.get(args.custom_field):
-            labels = {'__port__': str(args.port)}
-            if getattr(device, 'name', None):
-                labels['__meta_netbox_name'] = device.name
-            else:
-                labels['__meta_netbox_name'] = repr(device)
-            if device.tenant:
-                labels['__meta_netbox_tenant'] = device.tenant.slug
-                if device.tenant.group:
-                    labels['__meta_netbox_tenant_group'] = device.tenant.group.slug
-            if getattr(device, 'cluster', None):
-                labels['__meta_netbox_cluster'] = device.cluster.name
-            if getattr(device, 'asset_tag', None):
-                labels['__meta_netbox_asset_tag'] = device.asset_tag
-            if getattr(device, 'device_role', None):
-                labels['__meta_netbox_role'] = device.device_role.slug
-            if getattr(device, 'device_type', None):
-                labels['__meta_netbox_type'] = device.device_type.model
-            if getattr(device, 'rack', None):
-                labels['__meta_netbox_rack'] = device.rack.name
-            if getattr(device, 'site', None):
-                labels['__meta_netbox_pop'] = device.site.slug
-            if getattr(device, 'serial', None):
-                labels['__meta_netbox_serial'] = device.serial
-            if getattr(device, 'parent_device', None):
-                labels['__meta_netbox_parent'] = device.parent_device.name
-            if getattr(device, 'address', None):
-                labels['__meta_netbox_address'] = device.address
-            if getattr(device, 'description', None):
-                labels['__meta_netbox_description'] = device.description
```

Comment:
You removed a lot of labels I added (and I use). Please reintroduce them. If you don't want them, we have to find a way to keep the informations for the current users (maybe something configurable)?

But anyway, it should be motivated and isolated in a separated commit.